SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp

```diff
@@ -146,30 +162,71 @@ void USpatialSender::GainAuthorityThenAddComponent(USpatialActorChannel* Channel
 				PendingSubobjectAttachment);
 
-			EntityACL->ComponentWriteAcl.Add(Info->SchemaComponents[Type], AuthoritativeWorkerRequirementSet);
+			NewComponentIds.Add(ComponentId);
 		}
 	});
 
-	FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+	// If this worker is EntityACL authoritative, we can directly update the component IDs to gain authority over.
+	if (StaticComponentView->HasAuthority(EntityId, SpatialConstants::ENTITY_ACL_COMPONENT_ID))
+	{
+		const FClassInfo& ActorInfo = ClassInfoManager->GetOrCreateClassInfoByClass(Channel->Actor->GetClass());
+		const WorkerAttributeSet WorkerAttribute = { ActorInfo.WorkerType.ToString() };
+		const WorkerRequirementSet AuthoritativeWorkerRequirementSet = { WorkerAttribute };
+
+		EntityAcl* EntityACL = StaticComponentView->GetComponentData<EntityAcl>(Channel->GetEntityId());
+		for (auto& ComponentId : NewComponentIds)
+		{
+			EntityACL->ComponentWriteAcl.Add(ComponentId, AuthoritativeWorkerRequirementSet);
+		}
+
+		FWorkerComponentUpdate Update = EntityACL->CreateEntityAclUpdate();
+		Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
+	}
+
+	// Update the ComponentPresence component with the new component IDs. If this worker does not have EntityACL
+	// authority, this component is used to inform the enforcer of the component IDs to add to the EntityACL.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+	ComponentPresence* ComponentPresenceData = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	ComponentPresenceData->AddComponentIds(NewComponentIds);
+	FWorkerComponentUpdate Update = ComponentPresenceData->CreateComponentPresenceUpdate();
 	Connection->SendComponentUpdate(Channel->GetEntityId(), &Update);
 }
 
-void USpatialSender::SendRemoveComponent(Worker_EntityId EntityId, const FClassInfo& Info)
+void USpatialSender::SendRemoveComponentForClassInfo(Worker_EntityId EntityId, const FClassInfo& Info)
 {
+	TArray<Worker_ComponentId> ComponentsToRemove;
 	for (Worker_ComponentId SubobjectComponentId : Info.SchemaComponents)
 	{
 		if (SubobjectComponentId != SpatialConstants::INVALID_COMPONENT_ID)
 		{
 			NetDriver->Connection->SendRemoveComponent(EntityId, SubobjectComponentId);
+			ComponentsToRemove.Add(SubobjectComponentId);
 		}
 	}
 
+	// Update ComponentPresence.
+	check(StaticComponentView->HasAuthority(EntityId, SpatialConstants::COMPONENT_PRESENCE_COMPONENT_ID));
+	ComponentPresence* ComponentPresenceData = StaticComponentView->GetComponentData<ComponentPresence>(EntityId);
+	ComponentPresenceData->RemoveComponentIds(ComponentsToRemove);
+	FWorkerComponentUpdate Update = ComponentPresenceData->CreateComponentPresenceUpdate();
+	Connection->SendComponentUpdate(EntityId, &Update);
 	PackageMap->RemoveSubobject(FUnrealObjectRef(EntityId, Info.SchemaComponents[SCHEMA_Data]));
 }
 
+void USpatialSender::SendRemoveComponent(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
```

Comment:
Could this take an array of component ids? Then we could call it directly from the function above and prevent code duplication.

Reply:
Oh - like you do for `SendAddComponents`

Reply:
How often are we removing multiple components? If we want to have just one function instead of 2 then would it make more sense to keep the single remove case?

Reply:
I like the parity of both methods accepting arrays. Encapsulating the ComponentPresence update into it is nice and, where we do remove multiple components, we only want to do a single ComponentPresence update for all removals.

Reply:
Except we then have to make single element arrays every time we call it. The other way around we'd have to call it in a loop for multiple components. But which way around we do it should depend on what the actual expected use is.
Accepting a temporary array is generally a smell.
I'm in two minds about encapsulating the presence component into this stuff. I think it might be the right call but only because we are currently stuck with the sender and lack any kind of entity management abstraction.

Reply:
I think we have to pick between code duplication and temporary arrays because we have two cases to accommodate: where we want to remove one component from the presence, and where we want to remove many from the presence (but still in one component update). I'm in favour of reducing duplication, but if you feel strongly / can outline what you think it should be, I'm happy to change?

Reply:
I don't see duplication as an inherently bad thing, but allocating temporary arrays that normally contain 1 element mean we've likely done something wrong.
Especially as the update ordering between the component removed and the presence updated has some subtleties that we might have to deal with at some point (although hopefully won't and we aren't going to right now so this might be moot).
That said I think what I actually want here is a larger change than we are likely to put in this PR and probably applies to some other things as well. So I think do whichever for now and I'll approve the PR regardless.